src/gte_py/api/chain/utils.py

```
@@ -328,229 +254,562 @@ def make_web3(
+                        
+                        # Check if we still have too many pending after sync
+                        pending_after_sync = self.last_sent - self.last_confirmed
+                        
+                        if pending_after_sync < self.max_pending_window:
+                            self.logger.info(f"Window cleared after retry: confirmed={self.last_confirmed}, pending={pending_after_sync}")
+                            return
+                        
+                        self.logger.warning(f"Window still full after attempt: {pending_after_sync}/{self.max_pending_window} pending")
+                        
+                    except Exception as e:
+                        self.logger.error(f"Network error during window check attempt: {e}")
+                        continue
+                
+            # All retries exhausted and window still full
+            final_pending = self.last_sent - self.last_confirmed
+            raise RuntimeError(
+                f"Pending transaction window still full after {len(retry_delays)} retries: "
+                f"{final_pending}/{self.max_pending_window} transactions pending"
+            )
 
+    async def _sign_transaction(self, contract_func: "TypedContractFunction[Any]") -> SignedTransaction:
+        """
+        Sign a transaction with automatic nonce allocation.
+        
         Args:
-            nonce: The nonce of the transaction to cancel
+            contract_func: The contract function to execute
+            
+        Returns:
+            Signed transaction ready for submission
+        """
+        await self._check_pending_window()
+        
+        # Atomic nonce increment
+        async with self.nonce_lock:
+            nonce = self.last_sent
+            self.last_sent += 1
+        
+        try:
+            # Build transaction with required parameters
+            tx_params: TransactionDictType = {
+                "chainId": self.chain_id or 6342,
```

should throw error instead of defaulting to 6342

---

fixed in:

https://github.com/liquid-labs-inc/gte-python-sdk/pull/32/commits/3d50cc83c195cda774ea0cc141b5defd8a1468d6